rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -706,31 +668,106 @@ impl AssetClient {
+            let (dataset_update_tx, dataset_update_rx) = mpsc::unbounded_channel();
+            match DatasetClient::new(
+                dataset_definition.clone(),
+                dataset_update_rx,
+                &default_dataset_destinations,
+                asset_client.asset_ref.clone(),
+                asset_client.status.clone(),
+                asset_client.specification.clone(),
+                asset_client.device_specification.clone(),
+                asset_client.device_status.clone(),
+                asset_client.connector_context.clone(),
+            ) {
+                Ok(new_dataset_client) => {
+                    // insert the dataset client into the hashmap so we can handle updates
+                    asset_client.dataset_hashmap.insert(
+                        dataset_definition.name.clone(),
+                        (dataset_definition, dataset_update_tx),
+                    );
+
+                    // error is not possible since the receiving side of the channel is owned by the AssetClient
+                    let _ = asset_client.dataset_creation_tx.send(new_dataset_client);
+                }
+                Err(e) => {
+                    // Add the error to the status to be reported to ADR, and then continue to process
+                    // other datasets even if one isn't valid. Don't give this one to
+                    // the application since we can't forward data on it. If there's an update to the
+                    // definition, they'll get the create notification for it at that point if it's valid
+                    DatasetClient::update_dataset_status(
+                        &mut new_status,
+                        &dataset_definition.name,
+                        Err(e),
+                    );
+                    status_updated = true;
+                }
+            };
         }
 
-        // if there were any config errors, report them to the ADR service
-        asset_client
-            .report_dataset_config_errors(
-                dataset_config_errors,
-                specification_version,
-                "AssetClient::new dataset_destination(s)",
+        // if there were any config errors, report them to the ADR service together
+        if status_updated {
+            log::debug!(
+                "Reporting error asset status on new for {:?}",
+                asset_client.asset_ref
+            );
+            if let Err(e) = Self::internal_report_status(
+                new_status,
+                &asset_client.connector_context,
+                &asset_client.asset_ref,
+                &mut status_write_guard,
+                "AssetClient::new",
             )
-            .await;
+            .await
+            {
+                log::error!(
+                    "Failed to report error Asset status for new Asset {:?}: {e}",
+                    asset_client.asset_ref
+                );
+            }
+        }
+        drop(status_write_guard);
```

Comment:
Does this need to be dropped here if we're exiting the function?

Reply:
yeah, it's because the mutex lives on the AssetClient, so it's a little weird